Updating team workflow (#73)
2i2c-org/team-compass: practices/coordination.md

A deliverable is ready to work (and can thus be added to the Project Backlog) when it has the following properties (adapted from the [INVEST methodology](https://agileforall.com/new-to-agile-invest-in-good-user-stories/)).

- **Have a short title and description:** Deliverables should be glanceable and have enough context that the reader can quickly understand the scope.
- **Have one or more [user stories](https://www.atlassian.com/agile/project-management/user-stories):** User stories should define who benefits from the deliverable, and why they want it.
- **Have completion criteria:** We have clear completion criteria for this deliverable to denote it as “done”.
- **Have tasks to complete it**: Deliverables should have a set of tasks, which are actions needed to complete the deliverable.

#### How refinement happens

_Deliverable Refinement_ is the process of improving the scoping, context, and structure of our Deliverables issues so that they are ready for us to work on them. When a deliverable is created, it may not have all of the information needed to take the next step. Adding that information is the goal of Deliverable Refinement.

> All team members are expected to participate in deliverable refinement, though the more experience and higher-level you are, the more you should be contributing to this process.

Reviewer: ```suggestion
All team members are expected to participate in deliverable refinement, though the more experienced and higher-level you are, the more you should be contributing to this process.
```

We want to do this with high efficiency so that people can quickly get on the same page and focus efforts around decision-making and collaboration. We also wish to do it in a 💯 remote-friendly way since we are split across many different time-zones.

## Team development workflow

This section describes how our development team carries out its planning and day-to-day work.
Here's a quick summary:

> * [Deliverables](coordination:active-backlog) are discrete units of value that we bring to others. To be ready for work, they should be scoped properly, with enough information to complete them (e.g., a set of [tasks](coordination:tasks)).

Reviewer 2: I think the link (coordination:active-backlog) is incorrect here...

Author: Ah so in MyST Markdown, when being used with Sphinx, the markdown links behave like `:any:` roles in Sphinx. So this means that you can include either:

- A relative path to another file (e.g., `../index.md`)
- A URL (e.g. `https://google.com`)
- A cross-reference (e.g., `mylabel`)

In this case, I'm referencing the `coordination:active-backlog` section which is defined here:

https://github.com/2i2c-org/team-compass/pull/73/files#diff-b0c9a716af7c8f8c76b0af475a8b92ce614b96b340678951c855a7febad7654dR62

Reviewer 2: Yep, I quickly noted you wanted to cross-reference some other section... but my comment was triggered because I think the `coordination:active-backlog` reference is incorrect. Don't you want to reference the  `coordination: deliverables` section?

Author: ahhhh yep you are correct! I looked for a typo but hadn't considered that I'd just totally referenced the wrong section :-) thanks for spotting

Reviewer 2: I think you still did not fix this one in your latest commits, right?

Author: weird - you are correct! I just committed an update now

- {guilabel}`Needs Discussion/Refinement`: Deliverables that are high-priority but un-refined. Our goal should be having discussion and doing research in order to get these deliverables ready for work.
- {guilabel}`Ready to Work`: Deliverables that are well-scoped and have a clear path forward, and are thus ready to implement. As deliverables in {guilabel}`In progress` are completed, we should replace them with deliverables from this column. Generally speaking, deliverables near the top have higher priority than those at the bottom.
- {guilabel}`In progress`: Deliverables that we are currently working towards. This means that we should be generating [Tasks](coordination:tasks) in our [Active Development Board](https://github.com/orgs/2i2c-org/projects/5) to complete this deliverable (more on this below).
- {guilabel}`Blocked/Waiting`: Deliverables for which we are waiting for some action that is out of our team's immediate control.
- {guilabel}`Done`: Deliverables that have been completed. We should close these issues and celebrate the improvements that we have made!

:::{note}
In general, there should be only two or three deliverables per team member on this board.

> It should not become so heavily-populated that it is hard to keep track of deliverables!

Reviewer 2: You actually built a dev project board and an org project board... I guess you only want to mention the dev one here because of the dev-focused nature of this part of doc, am I right?

Author: yep correct - maybe we should add a "note" about the org one as well though. I am starting off just describing the "dev workflow" but we should generalize it over time.

Author: I've generalized this slightly and noted the locations of both project backlogs in the latest commit

We want to do this with high efficiency so that people can quickly get on the same page and focus efforts around decision-making and collaboration. We also wish to do it in a 💯 remote-friendly way since we are split across many different time-zones.

## Team development workflow

This section describes how our development team carries out its planning and day-to-day work.
Here's a quick summary:

> * [Deliverables](coordination:active-backlog) are discrete units of value that we bring to others. To be ready for work, they should be scoped properly, with enough information to complete them (e.g., a set of [tasks](coordination:tasks)).

Author: ```suggestion
* [Deliverables](coordination:deliverables) are discrete units of value that we bring to others. To be ready for work, they should be scoped properly, with enough information to complete them (e.g., a set of [tasks](coordination:tasks)).
```

* [Tasks](coordination:tasks) are concrete actions to take, and should be accomplishable in a single work session by one person.
* The [Activity Backlog](coordination:active-backlog) contains lists of **tasks** that we are currently working on. This defines 2i2c's current activity.

(coordination:project-backlog)=
### Project Backlogs

The Project Backlog contains all of the deliverables across our projects that we wish to work on quickly.
These are organized into a few columns, representing the _state_ of each deliverable.
Here is a common column structure:

- {guilabel}`Needs Discussion/Refinement`: Deliverables that are high-priority but un-refined. Our goal should be having discussion and doing research in order to get these deliverables ready for work.
- {guilabel}`Ready to Work`: Deliverables that are well-scoped and have a clear path forward, and are thus ready to implement. As deliverables in {guilabel}`In progress` are completed, we should replace them with deliverables from this column. Generally speaking, deliverables near the top have higher priority than those at the bottom.

> - {guilabel}`In progress`: Deliverables that we are currently working towards. This means that we should be generating [Tasks](coordination:tasks) in our [Active Development Board](https://github.com/orgs/2i2c-org/projects/5) to complete this deliverable (more on this below).

Author: ```suggestion
- {guilabel}`In progress`: Deliverables that we are currently working towards. This means that we should be generating [Tasks](coordination:tasks) in our [Activity Backlog](https://github.com/orgs/2i2c-org/projects/5) to complete this deliverable (more on this below).
```

- {guilabel}`Blocked/Waiting`: Deliverables for which we are waiting for some action that is out of our team's immediate control.
- {guilabel}`Done`: Deliverables that have been completed. We should close these issues and celebrate the improvements that we have made!

:::{note}
In general, there should be only two or three deliverables per team member, per column on this board.
It should not become so heavily-populated that it is hard to keep track of deliverables!
:::

There are two Project Backlogs at 2i2c:

- The [2i2c Development Projects Backlog](https://github.com/orgs/2i2c-org/projects/7) contains deliverables for our development-focused projects.

> - The [Organizatipnal Project Backlog](https://github.com/2i2c-org/meta/projects/6?fullscreen=true) contains deliverables that are organization-wide and not related to development.

Reviewer 3: ```suggestion
- The [Organizational Project Backlog](https://github.com/2i2c-org/meta/projects/6?fullscreen=true) contains deliverables that are organization-wide and not related to development.
```

> - The [Organizatipnal Project Backlog](https://github.com/2i2c-org/meta/projects/6?fullscreen=true) contains deliverables that are organization-wide and not related to development.

Reviewer 2: Maybe remove the full-screen from this last link? Or actually, add that to the previous link. The idea is to stay consistent.

Reviewer 2: @choldgraf I still see the discrepancy here, even after your :+1:

(coordination:deliverables)=
### Deliverables

Deliverables represent incremental amounts of value we can deliver to a particular stakeholder, and should be completable in a week or two. 

Most issues in our repositories are deliverables, in varying states of readiness. When a deliverable is first created, it may lack information, be improperly scoped, or have an unclear path to implementation. We improve this through _issue refinement_ (see below).

> A deliverable is ready to work (and can thus be added to the Project Backlog) when it has the following properties (adapted from the [INVEST methodology](https://agileforall.com/new-to-agile-invest-in-good-user-stories/)).

Author: Ah good point. My one concern is that if we automatically add anything w/ `Deliverable` on it, then it'll become extremely long (depending on how many deliverables we have across all of our repositories). What do you think about this?

> A deliverable is ready to work (and can thus be added to the Project Backlog) when it has the following properties (adapted from the [INVEST methodology](https://agileforall.com/new-to-agile-invest-in-good-user-stories/)).

Reviewer 3: Hmmm I'm a bit confused - are you commenting regarding automation associated with labels or the suggested change about "added" vs "advanced to that column"?

If we don't make this update, I'm considering how the column in the backlog about discussion is meant to be used. First add ready for work, and then it can transition back to needs discussion/refinement?

> A deliverable is ready to work (and can thus be added to the Project Backlog) when it has the following properties (adapted from the [INVEST methodology](https://agileforall.com/new-to-agile-invest-in-good-user-stories/)).

Reviewer 3: @choldgraf regarding automation of labels, I'm not sure. If we don't, it means we don't really keep track of our deliverables in one place as i imagined the project backlog would do.

That would be fine, but it raises a question about what deliverables and when to add/remove them from the project board.

> A deliverable is ready to work (and can thus be added to the Project Backlog) when it has the following properties (adapted from the [INVEST methodology](https://agileforall.com/new-to-agile-invest-in-good-user-stories/)).

Author: I think I had imagined something like:

|               | Need more info                     | Ready for Work       |
|---------------|------------------------------------|----------------------|
| Low Priority  | Stays in issues                    | Put in Ready to work |
| High Priority | Put in Needs Discussion/Refinement | Put in Ready to work |

However, I think you are suggesting a simpler model, like:

|               | Need more info                     | Ready for Work       |
|---------------|------------------------------------|----------------------|
| Low Priority  | Put in Needs Discussion/Refinement | Put in Ready to work |
| High Priority | Put in Needs Discussion/Refinement | Put in Ready to work |

Is that right? If so, I think I'm 👍 on trying your approach out, and seeing how it goes. We can use GitHub labels to tag priority high/low. What do you think?

> A deliverable is ready to work (and can thus be added to the Project Backlog) when it has the following properties (adapted from the [INVEST methodology](https://agileforall.com/new-to-agile-invest-in-good-user-stories/)).

Reviewer 3: Beautiful overview! Yes, that is what I meant and I now have a clear grasp on what I didn't before about what you meant!

I think both options make sense as long as they were a bit explicit, like both options now are.

I've just assumed it would be plausible to setup such automation, but I haven't done it. I think we should go for the non-automated approach until its proven viable. Looking into things I found some indication of viability but not by GitHub's official setup, but through https://github.com/philschatz/project-bot.

> A deliverable is ready to work (and can thus be added to the Project Backlog) when it has the following properties (adapted from the [INVEST methodology](https://agileforall.com/new-to-agile-invest-in-good-user-stories/)).

Author: Quick point on automation: I think the automation would be quite easy using ghapi (https://ghapi.fast.ai/) for example.

- **Have a short title and description:** Deliverables should be glanceable and have enough context that the reader can quickly understand the scope.
- **Have one or more [user stories](https://www.atlassian.com/agile/project-management/user-stories):** User stories should define who benefits from the deliverable, and why they want it.
- **Have completion criteria:** We have clear completion criteria for this deliverable to denote it as “done”.
- **Have tasks to complete it**: Deliverables should have a set of tasks, which are actions needed to complete the deliverable.

#### How refinement happens

_Deliverable Refinement_ is the process of improving the scoping, context, and structure of our Deliverables issues so that they are ready for us to work on them. When a deliverable is created, it may not have all of the information needed to take the next step. Adding that information is the goal of Deliverable Refinement.

> All team members are expected to participate in issue refinement, though the more experience and higher-level you are, the more you should be contributing to this process.

Author: good point - I'll call it "deliverable refinement" since I think that is a more precise term

The important thing is that we always have a list of high-quality deliverables ready to work towards.

(coordination:active-backlog)=
### Activity Backlog

The [Activity Backlog](https://github.com/orgs/2i2c-org/projects/5) contains the collection of **Tasks** that the team is currently working on.
Tasks are actions that are needed to accomplish some deliverable.
These tasks are generated from the deliverables on the Projects Backlog.
They define a “to do” list of tasks to complete on a day-to-day basis.

The Activity Backlog is broken down into these columns:

- {guilabel}`To Do` Tasks that are ready to be worked on. Roughly speaking, tasks higher on the list are of higher priority.

> - {guilabel}`Needs Discussion` Tasks that require some team discussion around _implementation_. If there are higher-level discussions about the deliverable itself, perhaps the deliverable is not yet ready to be worked on.

Author: good call - I will rename the column!

- {guilabel}`In progress` A task that a team member is currently working towards. When you move a task to In progress, indicate that you are working on it (either by adding your username to the card, or “assigning yourself” in the issue/PR associated with it.
- {guilabel}`Needs review` Tasks that have an implementation that require feedback from others. For example, reviews for pull requests.

> - {guilabel}`Done` Tasks that are complete! When you move a task here, make sure to update any relevant deliverables.

Author: huh weird - it is supposed to render like with a little "box" - maybe I need to release a new version of the sphinx book theme haha. I will try that
